addons/hr/models/hr_employee.py

```diff
@@ -644,20 +646,34 @@ def generate_random_barcode(self):
             employee.barcode = '041'+"".join(choice(digits) for i in range(9))
 
     def _get_tz(self):
-        # Finds the first valid timezone in his tz, his work hours tz,
-        #  the company calendar tz or UTC and returns it as a string
         self.ensure_one()
-        return self.tz or\
-               self.resource_calendar_id.tz or\
-               self.company_id.resource_calendar_id.tz or\
-               'UTC'
+        return self.tz  # required on `resource_resource`
 
     def _get_tz_batch(self):
         # Finds the first valid timezone in his tz, his work hours tz,
         #  the company calendar tz or UTC
         # Returns a dict {employee_id: tz}
         return {emp.id: emp._get_tz() for emp in self}
 
+    def _get_calendar_tz_batch(self, dt=None):
+        """ Find the effective schedule for the employee at dt -- based on their timezone.
+            Return dict {emp_id: schedule's timezone}"""
+        if not dt:
+            calendars = self._get_calendars()
+            return {emp_id: calendar.tz for emp_id, calendar in calendars.items()}
+
+        employees_by_tz = defaultdict(lambda: self.env['hr.employee'])
+        for employee in self:
+            employees_by_tz[employee._get_tz()] |= employee
+
+        employee_tzs = {}
+        for tz, employee_ids in employees_by_tz:
```

Comment:
:open_mouth: you're right -- I didn't notice because I also forgot to give the dt in `_get_day_start_and_day` so we're skipping this whole loop -- which matters only when we have contracts.
I skipped writing a test for when we specifically have a contract with a different schedule than the calendar because we're not editing anything in `hr_attendance_contract` but I guess this shows it's needed so I added it.